### PR TITLE
add dtypes to range dataset op. fix #33414

### LINKED ISSUE #33414: Allow to specify the type of the tf.data.Dataset.range function
**System information**
- TensorFlow version (you are using): *v2.0.0*
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**

[`tf.data.Dataset.range`](https://www.tensorflow.org/api_docs/python/tf/data/Dataset#range) return a dataset of `tf.dtypes.int64` tensors.

It could be nice if the `range` method accept a `dtype` argument (as many other TF methods) to choose the type of the dataset elements.

**Will this change the current api? How?**

This will change the API, adding an accepted `dtype` parameter to the `range` method.

**Who will benefit with this feature?**

Any person that uses the `Dataset.range` method to generate some index to use with `TensorArray.write` which only accept a `tf.dtypes.int32` tensor as its first argument.

**Any Other info.**

It could be possible to update `TensorArray.write` to let it accept `int32` or `int64` (or even `int8`) as its `index` argument.

### PR DESCRIPTION
**Pull request by ljwh**
Hi @jsimsa ,
This PR is just allow to specify the type of the tf.data.Dataset.range function, see issue #33414

I also add test code to range_dataset_op_test.cc and range_test.py.

### CODE REVIEW

**Review by jsimsa — CHANGES_REQUESTED** [review]
@ljwh thank you very much for the change, left some minor comments. 

You will also need to run the [API compatibility test](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/api/tests/api_compatibility_test.py) with `--update_goldens` flag set to True to reflect your change to the `tf.data.Dataset.range` API in the golden API files.

**Comment by ljwh** [comment]
Hi, @jsimsa  thank you for your comments, I updated the code and **please review**.

And I have a doubt, I add output_type param this way for compatible the previous code: 
```python
def range(*args, output_type=dtypes.int64):
```
so the call must be like this:
```python
# work
range(start, stop, step, output_type=tf.int64) 
# not work
range(start, stop, step, tf.int64) 
```
I don't think define interface this way is good, do you have any suggestions?

File: tensorflow/python/data/kernel_tests/range_test.py
Comment (jsimsa) [reply tensorflow/python/data/kernel_tests/range_test.py]: my suggestion was to avoid adding a new test `testOutputType` and instead parametrize each of the above test with different output types. After my https://github.com/tensorflow/tensorflow/commit/2a2c812ab2330c9aac33335f10679a346436acfb, you can do this by replacing:

```
@combinations.generate(test_base.default_test_combinations())
def testStop(self):
  dataset = dataset_ops.Dataset.range(5)
  ...
```

with:

```
@combinations.generate(combinations.times(
    test_base.default_test_combinations(),
    combinations.combine(output_type=[dtypes.int32, dtypes.int64, dtypes.float32, dtypes.float64])))
def testStop(self, output_type):
  dataset = dataset_ops.Dataset.range(5, output_type=output_type)
  ...
```

**Review by jsimsa — CHANGES_REQUESTED** [review]
Thank you @ljwh. As for your API comment, I do not see a better alternative to the one in which `output_type` needs to be specified through a keyword that would be backwards compatible at the same time.

Also, please resolve merge conflicts.

**Comment by ljwh** [comment]
Hi, @jsimsa , thank you for your guide, I updated the test code.

I am in doubt that why I could only pass the test code with eager mode while graph mode fail:
```python
# eager only, test passed
@combinations.generate(combinations.times(
    test_base.eager_only_combinations(),
    combinations.combine(output_type=[dtypes.int32, dtypes.int64, 
        dtypes.float32, dtypes.float64, dtypes.string])))

# eager and graph, test failed
@combinations.generate(combinations.times(
    test_base.default_test_combinations(),
    combinations.combine(output_type=[dtypes.int32, dtypes.int64, 
        dtypes.float32, dtypes.float64, dtypes.string])))
```

Could you please point me in the direction, thank you in advance.

File: tensorflow/python/data/kernel_tests/range_test.py
Comment (jsimsa) [inline tensorflow/python/data/kernel_tests/range_test.py]: Does it make sense to be using `dtypes.string` for all of these examples if the start / stop / step values are integers?

What would your expected use of `tf.data.Dataset.range` with `output_type=dtypes.string` look like?

**Comment by jsimsa** [comment]
What is the test failure with graph mode?

**Comment by ljwh** [comment]
Hi, @jsimsa , I found that:
```python
# next_ type defined as int64, 
# so the out_tensors is dtypes.int64
# and we can get output_dtype with   dataset()->output_dtypes()[0]
out_tensors->emplace_back(next_);
```
How to convert **next_** from **int64** to **output_dtype** here?

**Comment by jsimsa** [comment]
You should be able to do this:

```
...
Tensor result(dataset()->output_types()[0], dataset()->output_shapes()[0]);
switch (dataset()->output_types()[0]) {
#define HANDLE_TYPE(type)                                         \
  case DataTypeToEnum<type>::value: {                             \
    result->scalar<type>()() = static_cast<type>(next_);          \
    break;                                                        \
  }
TF_CALL_NUMBER_TYPES(HANDLE_TYPE);
#undef HANDLE_TYPE
  default:
    return errors::InvalidArgument("Unsupported data type: ", DataTypeString(dataset()->output_types()[0]));
}
out_tensors->push_back(std::move(result));
...
```

**Comment by ljwh** [comment]
Hi @jsimsa , thank you.

 I fixed the problem, please review.

File: tensorflow/python/data/kernel_tests/range_test.py
Comment (ljwh) [reply tensorflow/python/data/kernel_tests/range_test.py]: removed dtypes.string

**Comment by gbaned** [comment]
@ljwh thank you for your contribution, please sign CLA.

**Comment by ljwh** [comment]
@googlebot I fixed it.

**Review by jsimsa — DISMISSED** [review]

**Comment by gbaned** [comment]
@ljwh  Could you please check failed build errors? Thanks!

**Comment by ljwh** [comment]
Hi, @gbaned thank you for your reminder.

Hi, @jsimsa, I updated the code for backwards compatible to CI test, **please review**.

Looks like the CI would run after pr approved.

**Review by jsimsa — DISMISSED** [review]

**Comment by ljwh** [comment]
Hi, @jsimsa 
```python
  # Support in python3 but not support in python2(Syntax error)
  def range(*args, output_type=dtypes.int64):

  # So, I changed this func and considered the backwards compatible:
  def range(*args, **kwargs):
```
Also updated some code format for pylint, **Please review**

**Review by jsimsa — APPROVED** [review]

**Comment by ljwh** [comment]
hi @gbaned 
why ci does not run?

**Review by jsimsa — DISMISSED** [review]

**Comment by gbaned** [comment]
@ljwh  thank you for your contribution, please sign CLA.

**Comment by ljwh** [comment]
Hi, @jsimsa 
I updated the api golden files(by run api compatibility test).
Please review.

**Review by jsimsa — APPROVED** [review]

**Comment by ljwh** [comment]
Hi, @jsimsa 
CI build/test looks good, How to **trigger the copybara** ?

**Comment by jsimsa** [comment]
@gbaned could you please unblock this PR? thank you

**Comment by ljwh** [comment]
@gbaned could you please **trigger the copybara** ? thank you

**Comment by gbaned** [comment]
@jsimsa, @ljwh There is an automatic merge process that seems to be stalled for now. Thank you!

**Comment by ljwh** [comment]
Hi, @gbaned 
How is that going? Thank you

**Comment by ljwh** [comment]
@jsimsa @gbaned  thank you !

**Comment by jsimsa** [comment]
@gbaned what is this blocked on?

**Pull request merged**
